cleaner.py

```diff
@@ -0,0 +1,42 @@
+#! /usr/bin/env python3
+
+import logging
+import os
+import time
+
+import mailqueue
+
+
+def main():
+    cleanup_interval = int(os.environ.get('CLEANUP_INTERVAL', '300'))
+
+    # Cleanup implemented here is per-envelope, not per-submission. So it is possible that
+    # after the retention period the status of a submission might become invalid (e.g. if we've
+    # cleaned up one of the envelopes belonging to the submission, but there is another one still
+    # there).
+    #
+    # Making it always correct is slightly more complicated, so I consider this to be good enough
+    # for now.
+    #
+    # Because of the above we should set the retention period here to a higher value than what is
+    # communicated to customers (thus the factor of '2' below).
+    retention_period = int(os.environ.get('RETENTION_PERIOD', str(2 * 3 * 3600)))
+
+    setup_logging()
+
+    mq = mailqueue.MailQueue()
+    while True:
```

Comment:
What is the reasoning behind having a cleaner process? An alternative would be to make Sender removed envelopes that are no more needed. When it completes a sending attempt it can make a decision whether an envelope should be retried or not. In the latter case it can delete it?

Reply:
My understanding of the requirements was that we want to keep the status for the retention period for *all* submissions, not only those that are still pending.

E.g. if I submit an email, it gets delivered, I would still be able to query its status (and get back "delivered") within 3 hours of my submission.